articles/golang-advanced-programming-concepts/index.md

```diff
@@ -0,0 +1,174 @@
+
+*Go is a powerful statically-typed, open-source, and procedural programming language that is rapidly gaining popularity in the IT industry*. In the [first article](https://www.section.io/engineering-education/golang-part-1-introduction/), we looked at the history of Go, its purpose, and its installation, and in the [second article](https://www.section.io/engineering-education/golang-part-2-programming-basics/), we explored basic programming concepts such as variables, input/output, etc. In this article, we are going to go over advanced programming concepts such as functions, structures, and arrays. 
+
+### Functions
+A [function](https://www.tutorialspoint.com/computer_programming/computer_programming_functions.htm) can be defined as a set of reusable code that can be used to perform a particular task based on the requirements of the user. In other words, if there are a lot of repetitive tasks in the program, the user can create a function for that particular task and execute the function instead of writing the same piece of code multiple times. By creating functions, the user can avoid repetitive code and make the program easier to read and understand. By grouping similar tasks, the code becomes more [modular](https://en.wikipedia.org/wiki/Modular_programming) and more intuitive to other developers, thereby increasing productivity and efficiency in the team. Therefore, functions are extremely powerful, and it is generally considered good practice to use functions in your program. 
+
+The syntax for creating functions in Go is as follows:
+```
+func <name of function>(list of arguments) return type{
+	// block of reusable and repetitive code
+        return statement
+}
+```
+
+Let us create a function to add two numbers and return the result.
+
+```go
+package main
+
+import "fmt"
+
+func add_numbers(x int, y int) int {
+	return x + y
+}
+
+func main() {
+	var x = add_numbers(1, 2)
+        fmt.Println(x)
+}
+```
+
+The function `add_numbers()` takes 2 parameters: the two numbers to be added, and returns the sum of the two numbers. The two numbers are specified as `int` in the above example, but the parameters can be of any data type permissible in Go. The return type is also specified as int because the sum of two integers is also an integer.  
+
```

Comment:
You could add a small paragraph that specifies and explains different datatypes in Go. It would make it easier for beginners 😄

Reply:
Hi, @cyberShaw I have already done that in great detail in the previous articles. It would be redundant to do so here :) That's why I have linked the previous articles as prerequisites.

Reply:
Oh, right. Could you add a link to the particular article over there? 

For example, "The two numbers are specified as `int` in the above example, but the parameters can be of any data type permissible in Go. For more information on Go datatypes, please visit think link."

Reply:
Sure. I've made all the changes you've requested :)